### PR TITLE
Primitive caching

### PR DESCRIPTION
**Pull request by snadampal**
This PR is adding support for conv, matmul and inner product primitive caching. The feature is built on top of the existing primitive descriptor caching feature.

I have tested the below pytorch unit tests and there were no functional regressions observed.
test_ops.py
test_mkldnn.py
test_mkldnn_fusion.py
test_mkldnn_verbose.py
test_quantization.py

pytorch/benchmarks:  resnet50 with batchsize =32 showed ~4x improvement and the BERT_pytorch showed ~20% improvement.

### CODE REVIEW

**Comment by snadampal** [comment]
This feature has been submitted for PyTorch2.0 feature spec. Please let me know if I need to submit this change to a different branch. And also let me know what training specific tests I need to run for this. Thank you!

**Comment by snadampal** [comment]
Hi @XiaobingSuper , could you please comment on this PR? Please let me know what additional tests I need to run. this PR is targeted for PyTorch 2.0. Thank you!

**Comment by snadampal** [comment]
Hi, I see the ideep commit used on pytorch master as of today, but not sure which branch it corresponds to. Please help with merging this PR to the PT2.0 branch. Thank you!

>commit e7925bc7c260e6c4481ccb53b7d29c59a901a05d (HEAD, tag: pytorch-rls-v2.7.3, origin/ideep_pytorch)
Author: yanbing-j <yanbing.jiang@intel.com>
Date:   Mon Jan 16 16:13:44 2023 +0800
>
>    Switch mkl-dnn branch to dev-graph-beta-3-no-submod and Upgrade oneDNN to v2.7.3

**Comment by XiaobingSuper** [comment]
@snadampal , pytorch use this branch https://github.com/intel/ideep/commits/ideep_pytorch.

**Comment by jgong5** [comment]
cc @Xia-Weiwen @Guobing-Chen

File: include/ideep/operators/conv.hpp [hunk 1]
```diff
@@ -1354,6 +1354,7 @@ struct convolution_forward
       const dims& dilates,
       const dims& padding_l,
       const dims& padding_r,
+      const size_t weights_hashkey = 0, /*this is to check inplace weight updates*/
```
Comment (jgong5) [inline include/ideep/operators/conv.hpp:1357]: Why is the weight content related to the primitive cache logic?

File: include/ideep/operators/conv.hpp [hunk 2]
```diff
@@ -1395,10 +1396,11 @@ struct convolution_forward
...
         attr,
-        omp_get_max_threads());
-    return fetch_or_create(key, [&]() {
-      if (with_bias) {
-        return primitive_desc(
+        omp_get_max_threads(),
+        weights_hashkey);
+    dnnl::convolution_forward::primitive_desc pd;
+    if (with_bias) {
+      pd = primitive_desc(
```
Comment (jgong5) [inline include/ideep/operators/conv.hpp:1403]: Primitive descriptor creation takes time so it is not a good idea to create it every time. That's why we cache the pd.

**Review by jgong5 — CHANGES_REQUESTED** [review]
It is not a good idea to cache the primitive globally. The ideep cache is global and thread-local. Keeping primitives in the cache would bloat the memory (https://github.com/pytorch/pytorch/issues/29893). Also, oneDNN has the primitive cache internally. oneDNN primitive cache is global but not thread-local so more friendly to memory consumption. There is no need of maintaining another global cache inside ideep. The current logic is that we maintain the primitive descriptor in the cache since its creation takes time and then use it to create the primitive. Internally, oneDNN would do a cache lookup. Keeping a primitive in the ideep cache like this PR only saves the pd->primitive lookup time. I don't think it would save 4X for RN50 and 20% for BERT. Can you share more detailed profiling?

**Comment by snadampal** [comment]
Hi @jgong5, thanks for the feedback. Yes, I'm aware of the oneDNN primitive caching feature. But we can't use this with Arm Compute Library gemm kernels because they are non-immutable. Because the kernels are non-immutable, the ACL kernel objects are made runtime resources for oneDNN primitive, instead of objects of pd. and these resources get created as part of PyTorch primitive, so today they get created for every primitive create.

Moreover in TensorFlow we've seen that the operator level caching is more performant than the onednn level, hence TF operator caching is used even on x86. (https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/mkl/mkl_conv_ops.cc#L453)

> Keeping a primitive in the ideep cache like this PR only saves the pd->primitive lookup time. I don't think it would save 4X for RN50 and 20% for BERT. Can you share more detailed profiling?

As I explained above, oneDNN caching is not effective for ACL, and hence the primitive caching in this PR is bringing ~4x improvement for resnet50 and ~20% for Bert. This is how I tested.

```
git clone https://github.com/pytorch/benchmark.git

cd benchmark
sudo python3 install.py --continue_on_fail

export DNNL_DEFAULT_FPMATH_MODE=BF16
# for resnet50
 python3 run.py resnet50 -d cpu -m jit -t eval --use_cosine_similarity

# for bert
 python3 run.py BERT_pytorch -d cpu -m jit -t eval --use_cosine_similarity
```
btw, here is my oneDNN caching PR where we decided to not to enable it for ACL kernels.
https://github.com/oneapi-src/oneDNN/pull/1455

File: include/ideep/operators/conv.hpp (hunk 1 above)
Comment (snadampal) [reply include/ideep/operators/conv.hpp:1357]: Sorry, I could have added a note in the PR about this. This is required to allow inplace weight updates for the cached pds. We have seen this with some of the kernels where they reorder the weights for efficient gemm compute and cache the reordered weights internally for the subsequent pds with the assumption that the weights don't change. This will fail if the cached pd comes with updated weights. That's why I had to hash the weights to decide whether to reuse or recreate the primitive.

**Comment by snadampal** [comment]
I have raised the below PR on the ideep_pytorch branch
PR: https://github.com/intel/ideep/pull/162

File: include/ideep/operators/conv.hpp (hunk 1 above)
Comment (jgong5) [reply include/ideep/operators/conv.hpp:1357]: > We have seen this with some of the kernels where they reorder the weights for efficient gemm compute and cache the reordered weights internally for the subsequent pds with the assumption that the weights don't change

May I know what code you are talking about. Caching a reordered weight inside the kernel without knowing if it is a constant is by itself an issue. Also creating a hashkey with small portion of weight is not reliable.

Comment (jgong5) [reply include/ideep/operators/conv.hpp:1357]: For automatic weight prepacking in PyTorch and IPEX, we would first make sure that the weight is a constant before packing it into an optimal layout, e.g., the JIT script graph is frozen. Usually that happens in the inference path.

**Comment by jgong5** [comment]
> Moreover in TensorFlow we've seen that the operator level caching is more performant than the onednn level, hence TF operator caching is used even on x86. (https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/mkl/mkl_conv_ops.cc#L453)

Yes, I know that but it is due to historical reason when TF introduced the primitive cache in the framework before the primitive cache support is available in oneDNN and our TF folks didn't re-implement the code to use oneDNN primitive cache later. In PyTorch, we removed the primitive from the cache. We don't find obvious performance benefit of framework-level primitive cache vs. oneDNN primitive cache. As I explained, the saving is just the cache lookup cost in oneDNN, which is just about 1-2us per lookup. BTW, is TF primitive cache working on oneDNN/ACL now?

> But we can't use this with Arm Compute Library gemm kernels because they are non-immutable. Because the kernels are non-immutable, the ACL kernel objects are made runtime resources for oneDNN primitive, instead of objects of pd. and these resources get created as part of PyTorch primitive, so today they get created for every primitive create.

May I know what ACL runtime resources are maintained in oneDNN primitives right now? What are the memory usage impact of these resources in the primitives held in the cache?

> As I explained above, oneDNN caching is not effective for ACL, and hence the primitive caching in this PR is bringing ~4x improvement for resnet50 and ~20% for Bert. This is how I tested.

Thanks for the explanation why the perf gain is so big. I don't strongly object to caching the primitives from the framework side given this perf gain. But do you agree that for PyTorch, imperative path in particular, the final solution would be to let oneDNN handle the primitive caches after ACL kernel objects are made immutable, and the current solution is temporary and might have side-effect in memory usages? If so, I would suggest we do the following:
1. Only enable the primitive cache in ideep for ARM build and keep x86 build intact.
2. I still don't understand why adding weight hashkey is necessary. Suggest not add it. BTW, TF doesn't do that either, right?
3. Check the memory usage with benchmarks, especially with dynamic shapes. Communicate the result to the PyTorch community.

How does that sound to you?

cc @XiaobingSuper @Guobing-Chen to see if anything that I missed.

**Comment by snadampal** [comment]
> BTW, is TF primitive cache working on oneDNN/ACL now?

Yes, TF primitive caching is working with oneDNN/ACL .

>May I know what ACL runtime resources are maintained in oneDNN primitives right now? What are the memory usage impact of these resources in the primitives held in the cache?

ACL object is made as onednn resource. here is the conv primitive, the gemm object is created as part of the primitive:: create_resource()
https://github.com/oneapi-src/oneDNN/blob/master/src/cpu/aarch64/acl_gemm_convolution.hpp#L30
Memory impact wise it shouldn't be different from what it would cache it it was part of primitive descriptor and had it been cached via oneDNN.
If there are any runtime OOM concerns, we can always tune the `LRU_CACHE_CAPACITY` similar to `ONEDNN_PRIMITIVE_CACHE_CAPACITY` right?

> Only enable the primitive cache in ideep for ARM build and keep x86 build intact.

Yes, I agree, given that onednn caching is enabled for x86 already.

> I still don't understand why adding weight hashkey is necessary. Suggest not add it. BTW, TF doesn't do that either, right?

TF was doing it till TF2.10, here it is: https://github.com/tensorflow/tensorflow/blob/v2.10.0/tensorflow/core/kernels/mkl/mkl_conv_ops.cc#L488
 but TF2.11 onwards it was not required, because the later ACL versions support fixed format kernels , which made the weight reorders out side the base primitive. I can remove the weight hashkey from ideep when we move to oneDNN3.0.

> Check the memory usage with benchmarks, especially with dynamic shapes. Communicate the result to the PyTorch community.

Sure. Please let me know if you have any particular benchmarks in mind.

> How does that sound to you?

I'm perfectly fine with your suggestions. Thanks for the detailed review.!

**Comment by jgong5** [comment]
> TF was doing it till TF2.10, here it is: https://github.com/tensorflow/tensorflow/blob/v2.10.0/tensorflow/core/kernels/mkl/mkl_conv_ops.cc#L488
but TF2.11 onwards it was not required, because the later ACL versions support fixed format kernels , which made the weight reorders out side the base primitive. I can remove the weight hashkey from ideep when we move to oneDNN3.0.

But the way of hashing weight now is not sound. How can we guarantee we can use the cached reordered weight by just checking the leading bytes of it? Can we disable this reordered weight caching?

> ACL object is made as onednn resource. here is the conv primitive, the gemm object is created as part of the primitive:: create_resource()
> https://github.com/oneapi-src/oneDNN/blob/master/src/cpu/aarch64/acl_gemm_convolution.hpp#L30
> Memory impact wise it shouldn't be different from what it would cache it it was part of primitive descriptor and had it been cached via oneDNN.
> If there are any runtime OOM concerns, we can always tune the `LRU_CACHE_CAPACITY` similar to `ONEDNN_PRIMITIVE_CACHE_CAPACITY` right?

From the code you shared, ACL resource object would allocate and hold tensors inside it? oneDNN primitives on x86 won't do that - the tensors are managed and passed from the framework side. That means oneDNN primitives with ACL would consume a lot more memory?

> Sure. Please let me know if you have any particular benchmarks in mind.

Object detection models like MaskRCNN and NLP models like BERT that take dynamic seq lengths.

**Comment by snadampal** [comment]
Thanks @jgong5 . I'm making the changes aarch64 specific.

> But the way of hashing weight now is not sound. How can we guarantee we can use the cached reordered weight by just checking the leading bytes of it? Can we disable this reordered weight caching?

Doesn't look like a simple flag switch to disable the kernel level reordered weights caching for now. However, when we move to oneDNN3.0 with fixed format ACL kernels, the reorders are outside the kernel, then we don't need the weight hashkey. I'm checking if I can introduce some randomness in the key generation instead of just using the first 1k buffer.

> From the code you shared, ACL resource object would allocate and hold tensors inside it? oneDNN primitives on x86 won't do that - the tensors are managed and passed from the framework side. That means oneDNN primitives with ACL would consume a lot more memory?

Yes, you're right, today there is tensor allocation inside ACL and are part of the acl resource object. I've captured the memory foot print below.  Once we have the stateless kernels, we expect close to zero tensor allocations inside ACL, that should match x86 behavior and enable oneDNN level caching as well.

> Object detection models like MaskRCNN and NLP models like BERT that take dynamic seq lengths.

The memory foot print is ~3GB for [BERT_pytorch](https://github.com/pytorch/benchmark/tree/main/torchbenchmark/models/BERT_pytorch)and it is ~2.2GB for [resnet50](https://github.com/pytorch/benchmark/tree/main/torchbenchmark/models/resnet50) with batch size 1. and it goes to ~10GB for a batch size of 32.

**Comment by jgong5** [comment]
> The memory foot print is ~3GB for [BERT_pytorch](https://github.com/pytorch/benchmark/tree/main/torchbenchmark/models/BERT_pytorch)

Is this with dynamic seq length? What was the footprint before the change?

**Comment by snadampal** [comment]
Hi @jgong5,  I'm using the benchmark scripts from pytorch/benchmark repo. For the BERT_pytorch, dnnl traces show tensor of different shapes, but if I understand the scripts correctly, it looks like they are padding the input to the max sequence length, so probably not dynamic seq length.
https://github.com/pytorch/benchmark/blob/main/torchbenchmark/models/BERT_pytorch/bert_pytorch/dataset/dataset.py#L44

For the benchmarks from pytorch repo , I'm getting
```
BERT_pytorch: 
Memory: 10% increase in memory (2.89 GB on the previous release vs 3.2GB with this PR) 
latency: ~25% improvement

resnet50:
batchsize=1
Memory: ~30% increase in mefor resnet50 (1.69GB vs 2.2GB) for single stream and 
latency: ~2.67x improvement

batch size=32
Memory: doubled
latency: 4x improved
```

The plan is to cut down the memory by making the ACL kernels stateless in the future releases, so that either this PyTorch or oneDNN level caching gives the above latency improvements without the additional memory overhead.

btw, I posted update in the below PR to make the changes aarch64 specific, but keeping the same weights hashing logic.
https://github.com/intel/ideep/pull/162

**Comment by jgong5** [comment]
> btw, I posted update in the below PR to make the changes aarch64 specific, but keeping the same weights hashing logic.
> #162

How do you plan to handle the weight hashing logic? Another side effect of weight hashing is that it would bloat memory for training even if you can rely on the leading 1k elements as the key, right? Weights are updated for each iteration.

**Comment by jgong5** [comment]
@snadampal How about disable the LRU cache by default on ARM by changing the default capacity to 0?
https://github.com/intel/ideep/blob/e7925bc7c260e6c4481ccb53b7d29c59a901a05d/include/ideep/lru_cache.hpp#LL152-L152C76

Before this change, the ideep cache doesn't help perf on ARM anyway. So, we keep the default perf intact and avoid memory bloat by default. Users can set the LRU_CACHE_CAPACITY to a larger value to get better perf with this PR. Overall, this reduces the risks of the side effects introduced by this change (memory bloat and potential correctness issue due to weight caching). In the future, we can improve ACL kernels and make the default faster.

**Comment by jianan-gu** [comment]
> Hi @jgong5, I'm using the benchmark scripts from pytorch/benchmark repo. For the BERT_pytorch, dnnl traces show tensor of different shapes, but if I understand the scripts correctly, it looks like they are padding the input to the max sequence length, so probably not dynamic seq length. https://github.com/pytorch/benchmark/blob/main/torchbenchmark/models/BERT_pytorch/bert_pytorch/dataset/dataset.py#L44
> 
> For the benchmarks from pytorch repo , I'm getting
> 
> ```
> BERT_pytorch: 
> Memory: 10% increase in memory (2.89 GB on the previous release vs 3.2GB with this PR) 
> latency: ~25% improvement
> 
> resnet50:
> batchsize=1
> Memory: ~30% increase in mefor resnet50 (1.69GB vs 2.2GB) for single stream and 
> latency: ~2.67x improvement
> 
> batch size=32
> Memory: doubled
> latency: 4x improved
> ```
> 
> The plan is to cut down the memory by making the ACL kernels stateless in the future releases, so that either this PyTorch or oneDNN level caching gives the above latency improvements without the additional memory overhead.
> 
> btw, I posted update in the below PR to make the changes aarch64 specific, but keeping the same weights hashing logic. #162

Hi, 
For the benchmarking with dynamic shapes, would you like to try Huggingface examples? It provides a flag named _no_pad_to_max_length_ to make the dataset no pad for the sequence length. 
Test steps for run [question-answering](https://github.com/huggingface/transformers/tree/main/examples/pytorch/question-answering) task on Bert model with dynamic shapes:

After installing PyTorch,

    git clone https://github.com/huggingface/transformers
    cd transformers
    pip install -e ./     or    python setup.py install
    pip install -r examples/pytorch/question-answering/requirements.txt

Test: 

    cd transformers
    python  examples/pytorch/question-answering/run_qa.py --model_name_or_path   bert-large-uncased-whole-word-masking-finetuned-squad   --dataset_name squad   --do_eval   --no_pad_to_max_length   --output_dir ./tmp   --per_device_eval_batch_size 1 


[Optional]
- you can replace the model name with the one you find [here](https://huggingface.co/models?pipeline_tag=question-answering&sort=downloads&search=bert), e.g., " csarron/bert-base-uncased-squad-v1 "
- use --jit_mode_eval flag to enable pytorch jit path.
- use --no_cuda --bf16 to enable bf16 AMP on CPU

**Comment by snadampal** [comment]
Hi @jgong5 ,  Yes, without the PR,  the current LRU cache is of no use for aarch64. 

I already checked
`export LRU_CACHE_CAPACITY = 1 `(please note LRU_CACHE_CAPACITY=0 crashes, I think that needs to be fixed as a separate issue)
and `export LRU_CACHE_CAPACITY =1024`
and as expected, cache=1 brings the latencies and memory back to without the PR. 

Let me run the dynamic shape benchmarks and see how bad the memory overhead is. Thank for sharing the hugging face instructions.

**Comment by snadampal** [comment]
Hi @jgong5 , as expected, dynamic sequence length scenario didn't get any benefit from caching because there are no repetitive shapes, but the memory was doubled (~3GB vs ~7GB)

So, I'll go ahead with initializing LRU_CACHE_CAPACITY to 1 as default and increase it selectively for required inference use cases.

**Comment by jgong5** [comment]
> Hi @jgong5 , as expected, dynamic sequence length scenario didn't get any benefit from caching because there are no repetitive shapes, but the memory was doubled (~3GB vs ~7GB)
> 
> So, I'll go ahead with initializing LRU_CACHE_CAPACITY to 1 as default and increase it selectively for required inference use cases.

Sounds good to me.

**Comment by snadampal** [comment]
thanks, @jgong5 . updated the ideep_pytorch branch pr: https://github.com/intel/ideep/pull/162.

**Comment by snadampal** [comment]
closing this as it was already merged through PR: https://github.com/intel/ideep/pull/162
